compiler/src/dotty/tools/dotc/core/SymDenotations.scala

```diff
@@ -2132,9 +2130,24 @@ object SymDenotations {
           }
         case nil =>
           val directMembers = super.computeNPMembersNamed(name)
-          if (acc.exists) acc.union(directMembers.filterWithPredicate(!_.symbol.isAbsent()))
-          else directMembers
+          if !acc.exists then directMembers
+          else acc.union(directMembers.filterWithPredicate(!_.symbol.isAbsent())) match
+            case d: DenotUnion => dropStale(d)
+            case d => d
       }
+
+      def dropStale(d: DenotUnion): PreDenotation =
+        val compiledNow = d.filterWithPredicate(d =>
+          d.symbol.isDefinedInCurrentRun || d.symbol.associatedFile == null
+            // if a symbol does not have an associated file, assume it is defined
+            // in the current run anyway. This typically happens for pickling and
+            // from-tasty tests that generate a fresh symbol and then re-use it in the next run.
+          )
+        if compiledNow.exists then compiledNow
+        else
+          val youngest = d.aggregate(_.symbol.associatedFile.lastModified, _ max _)
```

Comment:
I'm strongly against any kind of merging logic that takes file modification date into account: there's many situations where this would go wrong, and it adds a new source of non-determinism to compilation. Instead, we can emit an error if overloaded definitions come from multiple files.

Reply:
That alone won't help. You still need the modification dates to get rid of junk in old classfiles. Otherwise any conflict would be an error which would be super annoying since often people do not even know how to clean out the output directory! So they are stuck with an error they can't fix.